File: codes/ABC/ABC165/ABC165_C.js
function main(input) {
    let [n, m, q] = input[0].trim().split(" ").map(e => parseInt(e))
    let [a,b,c,d] = [[],[],[],[]]
    for(let i = 0; i < q; i++) {
      [a[i],b[i],c[i],d[i]] = input[i+1].split(' ').map(e=>parseInt(e))
    }
    
    function dfs(arr, num, id) {
      arr.push(num)
      if(id < n-1) {
        let ret = 0
        for(let j = num; j <= m; j++) {
          let rep = dfs(arr.concat(), j, id+1)
          ret = Math.max.apply(null, [ret, rep])
        }
        return ret
      } else {
        let ret = 0
        for(let j = 0; j < q; j++) {
          if(arr[b[j]-1] - arr[a[j]-1] == c[j]) {
            ret += d[j]
          }
        }
        return ret
      }
    }
    
    let ar = []
    let ans = 0
    for(let i = 1; i <= m; i++) {
      let rep = dfs(ar.concat(), i, 0)
      ans = Math.max(ans, rep)
    }
    console.log(ans.toString())
  }
  
  main(require("fs").readFileSync("/dev/stdin", "utf8").trim().split('\n'))

File: codes/ABC/ABC165/ABC165_C.py
import copy

def dfs(ar, num, cnt):
  ar.append(num)
  if cnt < n-1:
    ans = 0
    for i in range(num, m+1):
      rep = dfs(copy.deepcopy(ar), i, cnt+1)
      ans = max([ans, rep])
    return ans
  else:
    ans = 0
    for i in range(q):
      if ar[abcd[i][1]-1] - ar[abcd[i][0]-1] == abcd[i][2]:
        ans += abcd[i][3]
    return ans

n,m,q = map(int, input().split())
abcd = [list(map(int, input().split())) for i in range(q)]

ans = 0
for i in range(1,m+1):
  a = []
  rep = dfs(copy.deepcopy(a), i, 0)
  ans = max([ans, rep])
print(ans)